Adding three new magnetic braking prescriptions to run_binary_extras.f and updated tidal-magnetic braking interaction

sgossage opened the pull request:
Added subroutines which calculate magnetic braking torque according to various prescriptions.

# Classical approach:
Classically, e.g., as in Rappaport, Verbunt, and Joss 1983, tidal locking is assumed and angular momentum is extracted directly from the orbital angular momentum by magnetic braking. To enable this method of magnetic braking for one of the three new prescriptions...

- In `inlist_project`:
```
    do_jdot_mb = .true.
    use_other_jdot_mb = .true.
```
- Then to select which magnetic braking prescription should be used, in `inlist1` set, e.g.,:

`    x_character_ctrl(1) = MB_option`

Where MB_option should be an integer:

1. MB_option = 'g18' (use [Garraffo+ 2018](https://ui.adsabs.harvard.edu/abs/2018ApJ...862...90G/abstract) torque)
2. MB_option = 'm15' (use [Matt+ 2015](https://ui.adsabs.harvard.edu/abs/2015ApJ...799L..23M/abstract) torque)
3. MB_option = 'carb' (use [Van & Ivanova 2019](https://ui.adsabs.harvard.edu/abs/2019ApJ...886L..31V/abstract) -- CARB torque)

Also, `do_jdot_ls = .false.`  should be set in `inlist_project` this case, as tidal locking is already assumed, making the simultaneous usage of LS coupling incompatible.

# Magnetic braking with tidal interactions:
In this case, we won't assume that tidal locking is applicable. Individual binary components will lose angular momentum via magnetic braking, tides will try to drive the system into synchronization, and tidal locking will only occur when/if the stellar rotation periods become comparable to their orbital period. Magnetic braking removes AM from individual stars, and tides try to either help or counteract this AM loss to bring the rotational periods in sync with the orbital period, either taking AM from the stars and adding it to the orbit, or taking AM from the orbit and putting it in the stars via LS coupling. 

A custom `jdot_ls` subroutine was written to account for AM lost via magnetic braking from the system. By default MESA calculates LS coupling as 

$$\delta J_{orb} = -[ ( \delta J_1 - \delta J_{w,1} \frac{ \dot M_{w,1} }{ \dot M_1 } ) + ( \delta J_2 - \delta J_{w,2} ) ]$$

where $\delta J_{orb}$ is the change in orbital angular momentum, $\delta J_i$ is the change in angular momentum of star $i$, and $ \delta J_{w,i}$ is the change due to winds. AM loss via winds is assumed to be lost from the system (not absorbed by the orbit), so that contribution is removed from $\delta J_i$, which could generically be written as,

$$\delta J_i = \delta J_{tides,i} + \delta J_{mb,i} + \delta J_{w,i}$$

(the extra factor $\frac{\dot M_{w,1}}{\dot M_1}$ means only the AM lost via winds not due to mass transfer is lost from the system). AM losses due to magnetic braking ($\delta J_{mb,i}$) should also be removed from the system (not absorbed by the orbit), so a similar modification to this equation needs to be done as was done for winds. The new equation should read:

$$\delta J_{orb} = -[(\delta J_1 - \delta J_{w,1}\frac{\dot M_{w,1}}{\dot M_1} - \delta J_{mb,1}) + (\delta J_2 - \delta J_{w,2} - \delta J_{mb,2})]$$

The new subroutine `my_jdot_ls` in `run_binary_extras.f` calculates this modified LS coupling torque.

## New magnetic braking term in tidal torque:
MESA does not normally consider sources/sinks of angular momentum when calculating tides within a given time step of evolution. The tidal torque calculation has been modified to account for magnetic braking (whose torque is stored as `s% extra_omegadot`) within a time step. This modification was made in the subroutine `my_sync_spin_to_orbit` in `run_binary_extras.f`. Essentially, the original change in stellar angular momentum due to tides, which read as

$$\delta J_{\rm orb-*} = J_{\rm orb} - J_{\rm *}$$

has been modified to

$$\delta J_{\rm orb-*'} = J_{\rm orb} - (J_{\rm *} + \delta J_{\rm *,mb})$$

where $\delta J_{\rm *,mb}$ is the change in angular momentum due to magnetic braking within a time step. Further notes on LS-tidal-magnetic braking interaction: [link](https://caramel-vulcanodon-91e.notion.site/Notes-on-tidal-magnetic-braking-interaction-in-MESA-b4743a0e62e34c38b0985f7cb1ddace4).

To enable this method of magnetic braking, set

- In `inlist_project`:
```
   do_jdot_mb = .false.
   do_jdot_ls = .true.`
   use_other_jdot_ls = .true.
```
and make sure that `do_tidal_sync = .true.`. Magnetic braking will be handled via `run_star_extras.f` in this case.

- In `inlist1` (and in `inlist2` if magnetic braking should also apply to star 2):
```
    use_other_torque = .true.
    x_character_ctrl(1) = MB_option
```

Where MB_option should be a string:

1. MB_option = 'g18' (use [Garraffo+ 2018](https://ui.adsabs.harvard.edu/abs/2018ApJ...862...90G/abstract) torque)
2. MB_option = 'm15' (use [Matt+ 2015](https://ui.adsabs.harvard.edu/abs/2015ApJ...799L..23M/abstract) torque)
3. MB_option = 'carb' (use [Van & Ivanova 2019](https://ui.adsabs.harvard.edu/abs/2019ApJ...886L..31V/abstract) -- CARB torque)
4. MB_option = 'rvj83' (use [Rappaport et al. 1983](https://ui.adsabs.harvard.edu/abs/1983ApJ...275..713R/abstract) torque)

When using this method, because the torque from magnetic braking scales with the rotation rate of the star, using `do_initial_orbit_sync = .true.` can lead to negligible magnetic braking torques for wide orbits. Using `do_initial_orbit_sync = .false.` may be preferred, with the individual binary components given some initial rotation rate that does not necessarily correspond to their orbital rotation period.

sgossage committed:
Added three new selectable prescriptions for magnetic braking.

sgossage committed:
Added three new selectable prescriptions for magnetic braking.

sgossage committed:
Added three new selectable prescriptions for magnetic braking.

sgossage committed:
Minor fix to indentation.

sgossage committed:
Another minor fix to indentation.

sgossage committed:
Added a check to set angular momentum loss to zero after removing AM from individual stars (when LS coupling is off) to ensure AM is not being taken from the orbital AM as well.

sgossage committed:
Fixed syntax error not -> .not. in various if statements.

mkruckow committed:
Detect PISN

Detect PISN and terminate MESA there to save the final profiles.

sgossage committed:
Added the ability to apply magnetic braking to individual binary components (via inlist1 and inlist2). In that case, the interaction between tides and magnetic braking should naturally drive the orbital evolution.

sgossage committed:
Updated tidal-magnetic braking coupled behavior in run_binary_extras.f

ZepeiX committed:
Update inlist1

ZepeiX committed:
Update inlist2

Commit:
Global definition of t_spindown was missing and has been added.

ZepeiX committed:
Merge pull request #15 from POSYDON-code/matthias_PISN

Detect PISN

Commit:
Changed the condition to trigger the AGB (Bloecker) wind. Previously this used MESA's s% have_done_TP, which is only set when convective shell overshoot is greater than 0. Now this uses a condition that checks that the He shell mass is not too large, and that central He has been depleted, to turn AGB winds on near when the TP-AGB wind phase begins.

Commit:
removed the < 10 msol check for when to use low mass winds. Temperature, He shell mass, and central He mass frac. decides which wind is used

Commit:
the Blocker wind evaluation now takes the max of the Reimers wind and the Blocker wind for a smooth transition

Commit:
Fixed erroneous definition of reimers_wind variable in eval_wind_for_scheme

Commit:
Removed if...else... surrounding the hot/cool wind scheme logic

Commit:
Merge remote-tracking branch 'upstream/seth_TPAGBwind_runstar' into main

syncing upstream TPAGBwind branch with magnetic braking fork to make it up to date

Commit:
Added check in extras_check_model to catch high mass loss rate convergence issues for low mass stars

Commit:
condition to check max mass transfer rate in extras_check_model should have been the absolute value. This has been corrected

Commit:
fixed error where n_conv_regions > 0 should be checked before attempting to access the cz_top_mass array. If there are no convection zones, the array would be empty causing a crash. Matt et al had this fix, and it is now propagated to the other prescriptions.

Commit:
fixed typo

Commit:
fixed typo

mkruckow commented:
This PR should be renewed to go into development branch instead of main.
I think it is bad, if we need to change other parameters to be able to select a certain MB_option. Hence I was thinking, whether a certain MB_option should change the other values and give a printout when the values got changed. This could be done in `extras_startup` or `extras_binary_startup`.

Commit:
fixed an issue with rvj_torque and carb_torque checking ocz_*_mass arrays when no conv. zone exists, causing an error

mkruckow commented on r11701/default_common_inlists/binary/src/run_star_extras.f:
```diff
@@ -1395,15 +1395,16 @@ subroutine carb_torque(id, ierr)
    if (s% n_conv_regions > 0) then
+
+      ocz_bot_mass = s% cz_bot_mass(i)
+      ocz_top_mass = s% cz_top_mass(i)
```
Btw. why do you define this mass variable, if you don't make use of it later?

sgossage replied:
thanks, those were superfluous and are now removed

sgossage committed:
Update code base with latest changes from v2 development (#1)

* add grid points for 2Zsun 0.2Zsun and 0.45Zsun

* adding 0.2Zsun 0.45Zsun ZAMS files

* adding 2Zsun ZAMS model

* fix CO-HMS_RLO 0.45Zsun csv files

* put the intrinsic 'Terminate because of overflowing initial model' to extras_binary_finish_step in run_binary_extras which seems to be checked anyway even if min_timestep_is_reached

* various prints

* added it to check_model and startup_model too in case it works

* calculate my own rl

* typo

* typo2

* typo2

* typo2

* typo2

* typo3

* typo3

* typo4

* only checking RL1

* cleaned for PR

* Update run_binary_extras.f

* implementing Matthias comments

* expansion single HMS grid to 0.1Msun

* surf tau between 1-100

* surf tau between 1-100

* omega_critlimit 0.95

* Update CO-HMS_RLO_yggdrasil.ini

add local work directory

* Update CO-HeMS_yggdrasil.ini

add local work directory

* Update HMS-HMS_yggdrasil.ini

add local work directory

* Update single_HMS_yggdrasil.ini

add local work directory

* Update single_HeMS_yggdrasil.ini

add local work directory

* Update CO-HMS_RLO_quest.ini

add local work directory

* Update CO-HeMS_quest.ini

add local work directory

* Update HMS-HMS_quest.ini

add local work directory

* Update single_HMS_quest.ini

add local work directory

* Update single_HeMS_quest.ini

add local work directory

* mesh_Pgas_div_P_exponent = 2 (DEF = 0)

* surf_avg_tau_min = 10d0

* add thermohaline mixing

* added x_logical_ctrl(6), to turn on magnetic braking during RLOF, only for He-Stars

* Update inlist1

Just changed the comment to make it more clear what this option does.

* Update inlist1

Chenged comment to be more clear what we do with this option.

* Update run_binary_extras.f

Changed again the comment to be more precise about wht this line of code does. We do not check for HMS stars here. This is done manually from teh inlist.

* Update inlist1

just added a phrase in teh comment

* Update run_star_extras.f

write energy error only once per model instead of every shell

* change boolean condition from default true to default false

* Dropping mesh_Pgas_div_P_exponent change for now in inlist1

* ...and max_allowed_nz in inlist1

* Dropping mesh_Pgas_div_P_exponent and max)allowed_nz changes for now in inlist2

* Seth hemb mltp mesh rerun (#28)

* Update inlist1

Add Seth's rerun changes for mtl+ and mesh control

* Update inlist2

Add Seth's rerun changes for mtl+ and mesh control

* Update run_star_extras.f

Add Z-burning power in luminosity comparison and request He burning to dominate.

* Update run_star_extras.f

Add Z-burning power in luminosity comparison and request He burning to dominate.

* Update run_star_extras.f

use power_nuc_burn instead of power_Z_burn (only available in newer MESA versions)

* Update run_star_extras.f

use power_nuc_burn instead of power_Z_burn (only available in newer MESA versions)

* Update inlist1

get old inlist before rerun6

* Update inlist2

Get old inlist before rerun6

* Update inlist1

get old inlist before rerun5

* Update inlist2

get old inlist before rerun5

* Update run_star_extras.f

%s L(1) -> L1
%s r(1) -> R1
remove post-MS check
use max of 1e-4 and other wind

* Unique wind prescription ids

same changes as in https://github.com/POSYDON-code/POSYDON-MESA-INLISTS/pull/35

* Update run_star_extras.f

exclude TPAGB stars going into LBV regime

* Update run_star_extras.f

exclude (stripped) He stars from LBV winds

* Update run_star_extras.f

lower surface X limit to 0.1

* Update run_star_extras.f

dedt_energy_eqn is now enabled after model 1, alongside lnPgas flag; convective_bdy_weight is set to 0 if center_gamma grows above 1 (core degeneracy is setting in)

* Update run_binary_extras.f

Added associated termination flags for cases of superthermal accretion and contact, and superthermal accretion with critical rotation

* Update inlist1

added MIST Lagrangian mesh controls, surf_avg_tau and _min, surf_w_div_w_crit_lim and _tol

* Update inlist2

inlist2 updated with the same changes as inlist1 -- note: both inlists also have thermohaline_coeff set to 17.5 (from 0)

* Update run_binary_extras.f

added return statements for several pre-existing termination conditions that did not have them (to stop multiple termination codes being reported)

* Update inlist1

removed fix_eps_grav_transition_to_grid, which was used in MIST but seems to cause convergence some issues in tests with our models

* Update inlist2

removed fix_eps_grav_transition_to_grid, which was used in MIST, but seems to cause some convergence issues in tests with our models

* LBV+thermo+dedt+noAgeLimit (#40)

* disable max_age

disable max_age to allow low mass stars to become WDs

* no TAMS stopping

Disable TAMS stopping of single, low mass stars. So, they can evolve to become WDs.

* Update inlist2

keep consistent with inlist1

---------

Co-authored-by: posydon-admin <ntqhzqeygbkwxbdnpo@bvhrs.com>
Co-authored-by: Jeff Andrews <jeffrey.andrews@ufl.edu>
Co-authored-by: ezapartas <ezapartas@gmail.com>
Co-authored-by: mkruckow <122798003+mkruckow@users.noreply.github.com>
Co-authored-by: tassos25 <tassos25@users.noreply.github.com>

sgossage committed:
removing superfluously defined variables

sgossage commented:
I have a new branch created a bit ago that simplifies inputs and that can be merged into development (addressing a comment from Matthias above as well).

I will update this PR with that branch (`seth_mb_lbv_thermo_dedt_noagelim`).

The new branch only requires that 

`do_jdot_ls = .true.` (optionally, as this is turned on at RLOF automatically)
`use_other_jdot_ls = .true.`

for the new MB style to be enabled. For the old style, just

`do_jdot_mb = .true.`

all in `inlist_project`. The option to set which MB prescription is still in `inlist1` and `inlist2`, set via string.
